File: 2024/min_add_to_make_parentheses_valid.cpp
#include <string>

//            C++
//     Time complexity - O (N)
//     Space complexity - O (1)

class MinimumAddToMakeParenthesesValid {
public:
    int minAddToMakeValid(const std::string& s) {
        int openParentheses = 0;
        int minAdd = 0;

        for (char c : s) {
            if (c == '(') {
                openParentheses++;
            } else if (c == ')') {
                if (openParentheses == 0) {
                    minAdd++;
                } else {
                    openParentheses--;
                }
            }
        }

        return minAdd + openParentheses;
    }
};

File: go/min_add_to_make_parentheses_valid.go
package main

//            Go
//     Time complexity - O (N)
//     Space complexity - O (1)

func minAddToMakeValid(s string) int {
	openParentheses := 0
	minAdd := 0

	for _, char := range s {
		if char == '(' {
			openParentheses++
		} else if char == ')' {
			if openParentheses == 0 {
				minAdd++
			} else {
				openParentheses--
			}
		}
	}

	return minAdd + openParentheses
}
